### PR TITLE
Add custom exporter for `kubectl top pod`

### PR DESCRIPTION
**Pull request by sed-i**
## Issue
We need more info to determine what resource limits to use in the load tests.

## Solution
Add a custom exporter for `kubectl top pod` so we know how much cpu, mem each pod of the bundle is utilizing.

TODO:
- [x] Create custom metrics exporter
- [x] Scrape the exporter
- [x] Add "top pod" panel to the dashboard

image

## Release Notes
Add custom exporter for `kubectl top pod`.

### CODE REVIEW

File: tests/load/gcp/cos-lite/pod_top_exporter.tpl.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    output = result.stdout.decode("utf-8").strip()
+
+    as_list = list(map(str.split, output.splitlines()))
+    # [['alertmanager-0', '1m', '51Mi'], ['catalogue-0', '1m', '38Mi'], ... ]
+
+    as_dict = {entry[0]: {"cpu": parse_quantity(entry[1]), "mem": parse_quantity(entry[2])} for entry in as_list}
+    # {'alertmanager-0': {'cpu': Decimal('0.056'), 'mem': Decimal('55574528.000')}, ...}
+
+    print(as_dict)
```
Comment (Abuelodelanada) [inline tests/load/gcp/cos-lite/pod_top_exporter.tpl.py:32]: Is this `print` necessary?

**Review by Abuelodelanada — APPROVED** [review]

File: tests/load/gcp/cos-lite/pod_top_exporter.tpl.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+from lightkube.utils.quantity import parse_quantity
+
+# Run with: `FLASK_APP=pod_top_exporter.py flask run`
+app = Flask(__name__)
+
+
+def get_top_pod() -> dict:
+    # Going through `current` directly because microk8s-kubectl.wrapper creates subprocesses which
+    # expect a login session
+    cmd = "/snap/microk8s/current/kubectl --kubeconfig /var/snap/microk8s/current/credentials/client.config top pod -n ${JUJU_MODEL_NAME} --no-headers".split()
```
Comment (rbarry82) [inline tests/load/gcp/cos-lite/pod_top_exporter.tpl.py:17]: This is way too brittle. Calling various `microk8s` commands differs between strictly confined and not. The amount of scraping needed to actually pull this in "properly" is [enormous](https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/values.yaml#L635), and it's honestly a PITA to get any of these out without something running inside the cluster, but there is a middle ground between "literally call `kubectl top pod`" and run in k8s.

That middle ground is:
`kubectl get --raw /apis/metrics.k8s.io/v1beta1 | jq`
```json
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "metrics.k8s.io/v1beta1",
  "resources": [
    {
      "name": "nodes",
      "singularName": "",
      "namespaced": false,
      "kind": "NodeMetrics",
      "verbs": [
        "get",
        "list"
      ]
    },
    {
      "name": "pods",
      "singularName": "",
      "namespaced": true,
      "kind": "PodMetrics",
      "verbs": [
        "get",
        "list"
      ]
    }
  ]
}
```

You can drill into `nodes` or `pods`.
For `kubectl get --raw "/apis/metrics.k8s.io/v1beta1/pods" | jq`, you'll get a list of all of the objects. Including metadata, like the namespace, which can be used to filter.
```json
    {                                                                                                                                                                                                                                               
      "metadata": {                                                                                                                                                                                                                                 
        "name": "coredns-597584b69b-tddzt",                                                                                                                                                                                                         
        "namespace": "kube-system",                                                                                                                                                                                                                 
        "creationTimestamp": "2023-02-26T02:08:55Z",                                                                                                                                                                                                
        "labels": {                                                                                                                                                                                                                                 
          "k8s-app": "kube-dns",                                                                                                                                                                                                                    
          "pod-template-hash": "597584b69b"                                                                                                                                                                                                         
        }                                                                                                                                                                                                                                           
      },                                                                                                                                                                                                                                            
      "timestamp": "2023-02-26T02:08:41Z",                                                                                                                                                                                                          
      "window": "17.843s",                                                                                                                                                                                                                          
      "containers": [                                                                                                                                                                                                                               
        {                                                                                                                                                                                                                                           
          "name": "coredns",                                                                                                                                                                                                                        
          "usage": {                                                                                                                                                                                                                                
            "cpu": "3491677n",                                                                                                                                                                                                                      
            "memory": "24656Ki"                                                                                                                                                                                                                     
          }                                                                                                                                                                                                                                         
        }                                                                                                                                                                                                                                           
      ]                                                                                                                                                                                                                                             
    }
```
All of the certs are in `/snap/microk8s/current/certs-beta/`. You can pretty much yank all of these in pure Python (*no subprocess*) with something like:

```bash
curl $KUBE_API/apis/apps/v1/deployments \
  --cacert /snap/microk8s/current/certs-beta/ca.crt \
  --cert /snap/microk8s/current/certs-beta/client.crt \
  --key /snap/microk8s/current/certs-beta/client.key
```

Yank the certs into the exporter on startup, and use `requests` or `urllib` or whatever you want to read it all and transform

**Review by rbarry82 — CHANGES_REQUESTED** [review]

**Comment by simskij** [comment]
https://github.com/kubernetes/kube-state-metrics should already contain the equivalent of `kubectl top` by default. I'm not sure whether https://github.com/simskij/kube-metrics-k8s-operator is in a working state or not, but could serve as a starting point (or even just deploying kube-state-metrics manually.

**Comment by rbarry82** [comment]
> https://github.com/kubernetes/kube-state-metrics should already contain the equivalent of `kubectl top` by default under the `containers_` namespace. I'm not sure whether https://github.com/simskij/kube-metrics-k8s-operator is in a working state or not, but could serve as a starting point (or even just deploying kube-state-metrics manually.

Unfortunately, those metrics (the entire `containers_` namespace, actually, and the baseline requirement for `top pod`) actually comes from [metrics-server](https://github.com/kubernetes-sigs/metrics-server), not `kube-state-metrics`. The closest we can get there is `kube_pod_container_resource_requests`, and that's pretty fuzzy.

The incoherence of the ~~philosophers~~ k8s monitoring. `node_exporter`, `metrics-server`, _and_ `kube-state-metrics` are required to get a full overview. The existing charm only provides 2 out of 3, but `metrics-server` could be a fair addition to get started.

For certs/tokens, I did at least remember that the `kubeconfig` has them, in which case what sort of microk8s we're running doesn't matter, and you can also snatch the token (if RBAC is enabled) out of it. Pulling them out of the kubeconfig (`.clusters.cluster.certificate-authority-data` and `.users.[i].user.token` are just base64ed and are otherwise usable) is probably a good way to go.

**Comment by simskij** [comment]
> > https://github.com/kubernetes/kube-state-metrics should already contain the equivalent of `kubectl top` by default under the `containers_` namespace. I'm not sure whether https://github.com/simskij/kube-metrics-k8s-operator is in a working state or not, but could serve as a starting point (or even just deploying kube-state-metrics manually.
> 
> Unfortunately, those metrics (the entire `containers_` namespace, actually, and the baseline requirement for `top pod`) actually comes from [metrics-server](https://github.com/kubernetes-sigs/metrics-server), not `kube-state-metrics`. The closest we can get there is `kube_pod_container_resource_requests`, and that's pretty fuzzy.
> 
> The incoherence of the ~philosophers~ k8s monitoring. `node_exporter`, `metrics-server`, _and_ `kube-state-metrics` are required to get a full overview. The existing charm only provides 2 out of 3, but `metrics-server` could be a fair addition to get started.
> 
> For certs/tokens, I did at least remember that the `kubeconfig` has them, in which case what sort of microk8s we're running doesn't matter, and you can also snatch the token (if RBAC is enabled) out of it. Pulling them out of the kubeconfig (`.clusters.cluster.certificate-authority-data` and `.users.[i].user.token` are just base64ed and are otherwise usable) is probably a good way to go.

Yeah, in that case I think adding metrics-server to the existing charm (or setting it up manually as part of the integration tests) feel far more tempting than both the PRs original suggestion or the alternative suggestion.

**Comment by rbarry82** [comment]
> Yeah, in that case I think adding metrics-server to the existing charm (or setting it up manually as part of the load tests) feel far more tempting than both the PRs original suggestion or the alternative suggestion.

The problem there, in some sense, is the [absolutely ridiculous](Yeah, in that case I think adding metrics-server to the existing charm (or setting it up manually as part of the load tests) feel far more tempting than both the PRs original suggestion or the alternative suggestion.) amount of scrape rules/targets/service-discovery needed to actually pull them all. Without adding SD to Prometheus, or setting up something else to scrape them out of there and push it to Prom, the raw JSON from the metrics API itself may be easier, as ugly as it is.

File: tests/load/gcp/cos-lite/pod_top_exporter.tpl.py (hunk 1 above)
Comment (sed-i) [reply tests/load/gcp/cos-lite/pod_top_exporter.tpl.py]: Removed.

File: tests/load/gcp/cos-lite/pod_top_exporter.tpl.py (hunk 2 above)
Comment (sed-i) [reply tests/load/gcp/cos-lite/pod_top_exporter.tpl.py:27]: This works as intended with classically confined microk8s, which is also pinned.
Will revisit later.
https://github.com/canonical/cos-lite-bundle/issues/70

**Pull request merged**